### DISCUSSION TITLE
Show "Read contract" on verified contract without requiring a third-party service token

### DISCUSSION THREAD

**Discussion by CleyFaye** [blockscout/blockscout#8698] [Q&A] [answered]
Hi,

We've been using blockscout for a while on a small private chain to provide an easily auditable registry built on geth.
In the former frontend, even without any "wallet" connected, a verified smartcontract could be queried easily.

In the new frontend, the "Read contract" tab is hidden unless a token from walletconnect is setup.
Having tried putting a random string in there, the "Read contract" (alongside the "Write contract") tabs show up, and the read operations works perfectly fine without any interaction with that third-party. However, the UI does display buttons that will initiate connection to said third party (and fail, because of an invalid token).

Would it be possible to enable the "Read contract" tab without such integration? Seeing that it actually works fine and do provide a useful interface to query the blockchain. Maybe some special hard coded value that would bypass this integration, or a flag to just display the "Read contract" tab without showing the "connect wallet" option that is useless there.

From a privacy perspective, initiating connection to some third party from an end-user browser for absolutely no reason is hard to justify, especially if the feature actually works without it.

**Comment by ekawibisono** [comment]
Can you provide a more detailed picture of your problem?

**Comment by CleyFaye** [comment]
Of course.

We use blockscout on a private ethereum chain (using geth) with a few SmartContract. These are fully verified on blockscout, and anyone opening their page can check the source code.
However, in the new frontend, there is no "read" tab that would allow anyone to manually run a view method or check a public property:

![image](https://github.com/blockscout/blockscout/assets/1065984/f5a3e4da-870a-4ee1-b7cc-8ebf09d0a2ee)

What I'm looking for is a way to actually have the "Read contract" tab as visible on the public ethereum blockscout explorer:

![image](https://github.com/blockscout/blockscout/assets/1065984/e19d2457-a725-4317-9602-e90172754859)

However, the only solution I found was to register on a third party service (walletconnect). Requiring a third party service at this point is not needed, as read operations do not require a wallet at all. In addition, potentially triggering request from end-user browsers toward this third party is not desirable since it is not needed; read operations works perfectly fine without a wallet connected (I tested that by putting a bogus key in `NEXT_PUBLIC_WALLET_CONNECT_PROJECT_ID`).

So my question is, is there a way to get the frontend to properly display the "Read contract" tab without adding that project key?

**Reply by ekawibisono** [reply]
Hi @CleyFaye ,
From my experience, in fact, to use the read and write contract function you are required to install the env `NEXT_PUBLIC_WALLET_CONNECT_PROJECT_ID`, because it validates your project so that it really runs correctly, whether it's a testnet or mainnet. Read it detail about this env : https://github.com/blockscout/frontend/blob/main/docs/ENVS.md
Read env about blockchain env writing contract : https://github.com/blockscout/frontend/blob/main/docs/ENVS.md#blockchain-interaction-writing-to-contract-etc

**Reply by CleyFaye** [reply]
I'm not sure I get what you mean. Setting up `NEXT_PUBLIC_WALLET_CONNECT_PROJECT_ID` is indeed stated in the documentation as a requirement to have the "Read contract" appear, but it is not technically required at all. There is nothing to validate from a third party to perform a read on a contract; you do not even need to use an existing "wallet" to do so.

Forcing the "Read contract" tab to appear (by putting a random string in it, completely unrelated to walletconnect have the immediate effect of the tab appearing and being *completely functional*. The only reason I find this solution suboptimal is that it also display both the "Write contract" and "Connect wallet" buttons; the "Write contract" part is useless without being connected to a wallet, and the "Connect wallet" button will initiate a network connection from the end-user to the connectwallet website, which is a privacy issue.

I'm asking for a solution that would alllow *only* the "Read contract" tab to show, as it actually work out of the box, without integration with walletconnect, allowing anyone (without a wallet) to run queries on the views of a smartcontract. Basically, the same thing that worked perfectly fine with the previous blockscout frontend.

**Reply by tom2drum** [reply, marked as answer]
@CleyFaye thanks for raising this issue. We are going to fix it in the next release or so - https://github.com/blockscout/frontend/issues/1313
